src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala

```diff
@@ -41,20 +41,40 @@ object DataTypes {
   val defaultLimitValue: Int = 10000
 
   /** Replaces timestamp represented as Long in predicates with one understood by the SQL */
-  private def replaceTimestampInPredicates(path: EntityPath, query: Query, metadataService: MetadataService)(
+  private def replaceTimestampInPredicatesAndSnapshot(path: EntityPath, query: Query, metadataService: MetadataService)(
       implicit executionContext: ExecutionContext
-  ): Future[Query] =
-    query.predicates.map { predicate =>
-      metadataService.getTableAttributesWithoutUpdatingCache(path).map { maybeAttributes =>
-        maybeAttributes.flatMap { attributes =>
-          attributes.find(_.name == dropAggregationPrefixes(predicate.field)).map {
-            case attribute if attribute.dataType == DataType.DateTime =>
-              predicate.copy(set = predicate.set.map(x => new Timestamp(x.toString.toLong).toString))
-            case _ => predicate
+  ): Future[Query] = {
+    val fetchedAttributes = metadataService.getTableAttributesWithoutUpdatingCache(path)
+    val newSnapshot =
+      fetchedAttributes.map { maybeAttributes =>
+        query.snapshot.flatMap { snapshot =>
+          maybeAttributes.flatMap { attributes =>
+            attributes.find(_.name == snapshot.field).map {
+              case attribute if attribute.dataType == DataType.DateTime =>
+                snapshot.copy(value = new Timestamp(snapshot.value.toString.toLong).toString)
+              case _ => snapshot
+            }
           }
-        }.toList
+        }
       }
-    }.sequence.map(pred => query.copy(predicates = pred.flatten))
+
+    val newPredicates =
+      fetchedAttributes.map { maybeAttributes =>
+        query.predicates.map { predicate =>
+          maybeAttributes.flatMap { attributes =>
+            attributes.find(_.name == dropAggregationPrefixes(predicate.field)).map {
+              case attribute if attribute.dataType == DataType.DateTime =>
+                predicate.copy(set = predicate.set.map(x => new Timestamp(x.toString.toLong).toString))
+              case _ => predicate
+            }
+          }.toList
```

Comment:
Sorry, I'm having quite a hard time figuring out what's happening here, due to the Future[Option[List]] nesting and the interaction with the query optional values.

Could add a type annotation for all the `val`s  help understanding how the code behaves?

Reply:
sure, done